sys/net/gnrc/transport_layer/tcp/gnrc_tcp.c

```diff
@@ -0,0 +1,630 @@
+    } else if(ret == -EADDRINUSE) {
+        DEBUG("gnrc_tcp.c : gnrc_tcp_connect() : local_port is already in use.\n");
+    }
+
+    /* Wait until a connection was established or closed */
+    while (ret >= 0 && tcb->state != CLOSED
+           && tcb->state != ESTABLISHED && tcb->state != CLOSE_WAIT) {
+        msg_receive(&msg);
+        switch (msg.type) {
+            case MSG_TYPE_CONNECTION_TIMEOUT:
+                DEBUG("gnrc_tcp.c : _gnrc_tcp_open() : CONNECTION_TIMEOUT\n");
+                _fsm(tcb, TIMEOUT_CONNECTION, NULL, NULL, 0);
+                ret = -ETIMEDOUT;
+                break;
+
+            case MSG_TYPE_NOTIFY_USER:
+                DEBUG("gnrc_tcp.c : _gnrc_tcp_open() : MSG_TYPE_NOTIFY_USER\n");
+                break;
+
+            default:
+                DEBUG("gnrc_tcp.c : _gnrc_tcp_open() : other message type\n");
+        }
+    }
+
+    /* Cleanup */
+    if (ret != -ETIMEDOUT) {
+      xtimer_remove(&connection_timeout_timer);
+    }
+    if (tcb->state == CLOSED && ret == 0) {
+        ret = -ECONNREFUSED;
+    }
+    tcb->owner = KERNEL_PID_UNDEF;
+    mutex_unlock(&(tcb->function_lock));
+    return ret;
+}
+
+/* External GNRC_TCP API */
+int8_t gnrc_tcp_init(void)
+{
+    /* Guard: Check if thread is already running */
+    if (_gnrc_tcp_pid != KERNEL_PID_UNDEF) {
+        return -1;
+    }
+
+    /* Initialize Mutex for linked-list synchronization */
+    mutex_init(&(_list_gnrc_tcp_tcb_lock));
+
+    /* Initialize Linked-List for connection storage */
+    _list_gnrc_tcp_tcb_head = NULL;
+
+    /* Initialize receive buffers */
+    _rcvbuf_init();
+
+    /* Start TCP processing loop */
+    return thread_create(_stack, sizeof(_stack), GNRC_TCP_PRIO, 0, _event_loop, NULL, "gnrc_tcp");
+}
+
+int8_t gnrc_tcp_tcb_init(gnrc_tcp_tcb_t* tcb, const uint8_t address_family,
+                         void * const local_addr, uint16_t * const local_port,
+                         void * const peer_addr, uint16_t * const peer_port)
+{
+    /* Check Address Family */
+    switch (address_family) {
+#ifdef MODULE_GNRC_IPV6
+        case AF_INET6:
+            tcb->address_family = AF_INET6;
+            break;
+#endif
+        default:
+            DEBUG("gnrc_tcp.c : gnrc_tcp_tcb_init() : address_family not supported\n");
+            return -EAFNOSUPPORT;
+    }
+
+    /* Assign Address Buffers */
+    tcb->local_addr = local_addr;
+    tcb->peer_addr = peer_addr;
+    tcb->local_port = local_port;
+    tcb->peer_port = peer_port;
+
+    /* Initialize Transmission Control Block variables */
+    switch (address_family) {
+#ifdef MODULE_GNRC_IPV6
+        case AF_INET6:
+            ipv6_addr_set_unspecified((ipv6_addr_t *) tcb->local_addr);
+            ipv6_addr_set_unspecified((ipv6_addr_t *) tcb->peer_addr);
+            break;
+#endif
+    }
+    *(tcb->local_port) = GNRC_TCP_PORT_UNSPEC;
+    *(tcb->peer_port) = GNRC_TCP_PORT_UNSPEC;
```

Comment:
Yes because the address and port fields should not contain here target address info here. Thats that the open calls are for. At this Point the port and address are initialized. Later called code relies on that.